File: src/steam/games.rs
use std::env;

use anyhow::{Context, Result};
use chrono::{DateTime, TimeZone, Utc};
use reqwest::Client;
use serde::{Deserialize, Serialize};

use super::{achievements, STEAM_ID, STEAM_TOKEN};

#[derive(Debug, PartialEq, Deserialize, Serialize, Clone)]
pub struct MainSteamResponse {
    pub response: SteamResponse,
}

#[derive(Debug, PartialEq, Deserialize, Serialize, Clone)]
pub struct SteamResponse {
    pub games: Vec<RawGame>,
}

#[derive(Debug, PartialEq, Deserialize, Serialize, Clone)]
pub struct RawGame {
    pub name: String,
    pub appid: u32,
    pub img_icon_url: String,
    pub rtime_last_played: i64,
    pub playtime_forever: u32,
}

#[derive(Debug, PartialEq, Deserialize, Serialize, Clone)]
pub struct Game {
    pub name: String,
    pub app_id: u32,
    pub url: String,
    pub icon_url: String,
    pub header_url: String,
    pub library_url: Option<String>,
    pub playtime_forever: u32,
    pub rtime_last_played: DateTime<Utc>,
    pub achievement_progress: f32,
}

pub async fn fetch_recently_played(client: &Client) -> Result<Vec<Game>> {
    let resp: reqwest::Response = client
        .get("https://api.steampowered.com/IPlayerService/GetOwnedGames/v1/")
        .query(&[
            (
                "key",
                env::var(STEAM_TOKEN).context("fetching steam token env var failed")?,
            ),
            (
                "steamid",
                env::var(STEAM_ID).context("fetching steam id env var failed")?,
            ),
            ("include_appinfo", String::from("true")),
            ("format", String::from("json")),
        ])
        .send()
        .await
        .context("sending request for recent games failed")?;
    let resp_text = resp
        .text()
        .await
        .context("getting raw response text failed")?;
    let mut data: MainSteamResponse = serde_json::from_str(&resp_text).context(format!(
        "reading json failed from request to get recent games: response: {}",
        resp_text
    ))?;
    data.response
        .games
        .sort_by(|a, b| b.rtime_last_played.cmp(&a.rtime_last_played));
    let mut games: Vec<Game> = vec![];
    for game in data.response.games.iter_mut() {
        let library_url = format!(
                "https://shared.akamai.steamstatic.com/store_item_assets/steam/apps/{}/library_600x900.jpg",
                &game.appid,
            );
        let library_url_exists = client
            .get(&library_url)
            .send()
            .await
            .context(format!("checking library url for {} failed", &game.name))?
            .status()
            == 200;
        let achievements_data = achievements::fetch_game_achievements(game.appid, client).await?;
        let mut completed_achievements = 0;
        achievements_data.iter().for_each(|a| {
            if a.achieved == 1 {
                completed_achievements += 1
            }
        });
        games.push(Game {
            name: game.name.to_owned(),
            url: format!("https://store.steampowered.com/app/{}/", &game.appid),
            icon_url: format!(
                "https://media.steampowered.com/steamcommunity/public/images/apps/{}/{}.jpg",
                &game.appid, game.img_icon_url
            ),
            library_url: if library_url_exists {
                Some(library_url)
            } else {
                None
            },
            header_url: format!(
                "https://shared.akamai.steamstatic.com/store_item_assets/steam/apps/{}/header.jpg",
                &game.appid,
            ),
            app_id: game.appid,
            playtime_forever: game.playtime_forever,
            rtime_last_played: Utc.timestamp_opt(game.rtime_last_played, 0).unwrap(),
            achievement_progress: achievements_data.len() as f32 / completed_achievements as f32,
        })
    }
    Ok(games)
}

File: internal/api/steam/games.go
package steam

import (
	"cmp"
	"encoding/json"
	"fmt"
	"image/jpeg"
	"net/http"
	"net/url"
	"sort"
	"strings"
	"time"

	"github.com/redis/go-redis/v9"
	"go.mattglei.ch/lcp/internal/api"
	"go.mattglei.ch/lcp/internal/cache"
	"go.mattglei.ch/lcp/internal/images"
	"go.mattglei.ch/lcp/internal/secrets"
	"go.mattglei.ch/lcp/pkg/lcp"
)

const storeItemAssetsBaseURL = "https://shared.akamai.steamstatic.com/store_item_assets/"

type ownedGamesResponse struct {
	Response struct {
		Games []struct {
			Name                     string `json:"name"`
			AppID                    int    `json:"appid"`
			LastPlayed               int64  `json:"rtime_last_played"`
			ImgIconURL               string `json:"img_icon_url"`
			PlaytimeForever          int    `json:"playtime_forever"`
			HasCommunityVisibleStats bool   `json:"has_community_visible_stats"`
		} `json:"games"`
	} `json:"response"`
}

type storeItemsResponse struct {
	Response struct {
		StoreItems []struct {
			AppID  int             `json:"appid"`
			Assets storeItemAssets `json:"assets"`
		} `json:"store_items"`
	} `json:"response"`
}

// storeItemAssets holds the versioned asset paths Steam returns for a store item. Newer apps
// have a content hash in the filename (e.g. "803663.../header.jpg") while older apps return a
// bare filename, so the unhashed convenience path can 404. AssetURLFormat is a template like
// "steam/apps/4121170/${FILENAME}?t=1781141362" that we fill in per asset.
type storeItemAssets struct {
	AssetURLFormat string `json:"asset_url_format"`
	Header         string `json:"header"`
	LibraryHero    string `json:"library_hero"`
}

func (a storeItemAssets) assetURL(appID int, filename string) string {
	format := cmp.Or(a.AssetURLFormat, fmt.Sprintf("steam/apps/%d/${FILENAME}", appID))
	return storeItemAssetsBaseURL + strings.Replace(format, "${FILENAME}", filename, 1)
}

func fetchStoreItemAssets(client *http.Client, appIDs []int) (map[int]storeItemAssets, error) {
	ids := make([]map[string]int, len(appIDs))
	for i, id := range appIDs {
		ids[i] = map[string]int{"appid": id}
	}
	inputJSON, err := json.Marshal(map[string]any{
		"ids": ids,
		"context": map[string]string{
			"language":     "english",
			"country_code": "US",
		},
		"data_request": map[string]bool{
			"include_assets": true,
		},
	})
	if err != nil {
		return nil, fmt.Errorf("encoding store items input json: %w", err)
	}

	params := url.Values{"input_json": {string(inputJSON)}}
	req, err := http.NewRequest(
		http.MethodGet,
		"https://api.steampowered.com/IStoreBrowseService/GetItems/v1/?"+params.Encode(),
		nil,
	)
	if err != nil {
		return nil, fmt.Errorf("creating request for steam store items: %w", err)
	}
	storeItems, err := api.RequestJSON[storeItemsResponse](client, req, logger())
	if err != nil {
		return nil, fmt.Errorf("sending request for store items: %w", err)
	}

	assets := make(map[int]storeItemAssets, len(storeItems.Response.StoreItems))
	for _, item := range storeItems.Response.StoreItems {
		assets[item.AppID] = item.Assets
	}
	return assets, nil
}

func fetchRecentlyPlayedGames(client *http.Client, rdb *redis.Client) ([]lcp.SteamGame, error) {
	params := url.Values{
		"key":             {secrets.ENV.SteamKey},
		"steamid":         {secrets.ENV.SteamID},
		"include_appinfo": {"true"},
	}
	req, err := http.NewRequest(
		http.MethodGet,
		"https://api.steampowered.com/IPlayerService/GetOwnedGames/v1/?"+params.Encode(),
		nil,
	)
	if err != nil {
		return nil, fmt.Errorf("creating request for steam API owned games: %w", err)
	}
	ownedGames, err := api.RequestJSON[ownedGamesResponse](client, req, logger())
	if err != nil {
		return nil, fmt.Errorf("sending request for owned games: %w", err)
	}

	sort.Slice(ownedGames.Response.Games, func(i, j int) bool {
		return ownedGames.Response.Games[j].LastPlayed < ownedGames.Response.Games[i].LastPlayed
	})

	if len(ownedGames.Response.Games) < 6 {
		return nil, cache.ErrSteamOwnedGamesEmpty
	}

	topGames := ownedGames.Response.Games[:10]
	appIDs := make([]int, len(topGames))
	for i, g := range topGames {
		appIDs[i] = g.AppID
	}
	assets, err := fetchStoreItemAssets(client, appIDs)
	if err != nil {
		return nil, err
	}

	var games []lcp.SteamGame
	for _, g := range topGames {
		gameAssets := assets[g.AppID]
		game := lcp.SteamGame{
			Name:  g.Name,
			AppID: g.AppID,
			IconURL: fmt.Sprintf(
				"https://media.steampowered.com/steamcommunity/public/images/apps/%d/%s.jpg",
				g.AppID,
				g.ImgIconURL,
			),
			RTimeLastPlayed:    time.Unix(g.LastPlayed, 0),
			PlaytimeForever:    g.PlaytimeForever,
			URL:                fmt.Sprintf("https://store.steampowered.com/app/%d/", g.AppID),
			HeaderURL:          gameAssets.assetURL(g.AppID, cmp.Or(gameAssets.Header, "header.jpg")),
			LibraryHeroURL:     gameAssets.assetURL(g.AppID, cmp.Or(gameAssets.LibraryHero, "library_hero.jpg")),
			LibraryHeroLogoURL: gameAssets.assetURL(g.AppID, "logo.png"),
		}

		if g.HasCommunityVisibleStats {
			achievementPercentage, err := fetchAchievementsPercentage(client, g.AppID)
			if err != nil {
				return nil, err
			}
			game.AchievementProgress = achievementPercentage
		}

		headerBlurHash, err := images.BlurHash(client, rdb, game.HeaderURL, jpeg.Decode, logger())
		if err != nil {
			logger().Warn().Err(err).Int("app_id", g.AppID).Msg("failed to generate header blurhash")
		} else {
			game.HeaderBlurHash = headerBlurHash
		}

		games = append(games, game)
	}

	return games, nil
}
